lib/Raven/Client.php

```diff
@@ -233,35 +238,49 @@ class Client
      */
     protected $_shutdown_function_has_been_set;
 
-    public function __construct($options_or_dsn = null, $options = [])
+    /**
+     * @var Configuration The client configuration
+     */
+    protected $config;
+
+    /**
+     * Constructor.
+     *
+     * @param Configuration $config The client configuration
+     */
+    public function __construct(Configuration $config)
```

Comment:
>There's no reason for class factories

Class factories allows users to choose how an object is created, instead of relying on an internal behaviour of the class that's going to use the instances. For example, the processors are now simply created by doing a `new $processorClass(...)` and this is a big limit because if someone wants to have custom parameters in the constructor, access other framework services or whatever else he can't, at least without subclassing the `createProcessors` method. Instead, with a factory we could support a simple factory that just instantiates the processor, and we could support more complex factories in the framework-related packages that for example uses the service container to instantiate the processors. The same goes for the transport (and this is just an example). I understand that you want to keep the library as simple as possible, but introducing factories and refactoring the code in this way imho is not making it overly complex, is just about making it more extensible and better organized.

>e.g. how is that any better than $client = new Client(...)->setX();

It's better because with factories (and this is just an example) we can lazy-load the transport and processors. It's better because we won't restrict anymore how a given transport/processor must be created because we will shift this to a dedicated object that can be customized by the user to support dependency injection. I really can only see advantages in doing this. Let's be honest, who is going to use this library is a programmer, and I highly doubt that he will complain if he has to pass more arguments to the constructor if it means that he has more flexibility